src/AbstractTwigCallable.php

```diff
@@ -68,8 +74,23 @@ public function needsContext(): bool
         return $this->options['needs_context'];
     }
 
+    public function withDynamicArguments(string $name, string $dynamicName, array $arguments): self
+    {
+        $new = clone $this;
+        $new->name = $name;
+        $new->dynamicName = $dynamicName;
+        $new->arguments = $arguments;
+
+        return $new;
+    }
+
+    /**
+     * @deprecated since Twig 3.12, use withDynamicArguments() instead
+     */
     public function setArguments(array $arguments): void
```

Comment:
This is the only method that makes Twig callables mutable. And as demonstrated by the new tests, this can lead to bad behaviors.